### PR TITLE
Handle the error when user selects non-existing columns in the table

### LINKED ISSUE #70: handle error when user selects column that is not in the table
Query to reproduce:
```sql
create schema schema_name;
create table schema_name.table_name (column_in_table smallint);
select column_not_in_table from schema_name.table_name;
```

Expected result in `psql`:
```
ERROR: column "column_not_in_table" does not exist
```
additionally it could say the error line in query after the error message, like:
```
LINE 1: select col1 from empty;
```

Actual result:
```
select column_not_in_table from schema_name.table_name;
 column_not_in_table
---------------------
(0 rows)
```

The module to have a look is `storage/src/frontend/mod.rs`.
There is a method:
```
pub fn select_all_from(
        &mut self,
        schema_name: &str,
        table_name: &str,
        columns: Vec<String>,
    ) -> SystemResult<Result<Projection, OperationOnTableError>>
```
Before performing a query it reads table columns with `self.table_columns(schema_name, table_name)?`
so we have to add the check if column(s) in the table
also `sql_engine` module and `QueryResultMapper` in `node` module have to be updated respectively

### PR DESCRIPTION
**Pull request by silathdiir**
Fixes #70 

Tests with unit-tests and `psql` as below.

```
psql (12.1, server 0.0.0)
Type "help" for help.

asongala=> create schema schema_name;
CREATE SCHEMA
asongala=> create table schema_name.table_name (column_in_table smallint);
CREATE TABLE
asongala=> select column_not_in_table from schema_name.table_name;
ERROR:  column "column_not_in_table" does not exist
```

### CODE REVIEW

**Comment by silathdiir** [comment]
@alex-dukhno Please help review the code. Thanks.

**Comment by silathdiir** [comment]
Hi @alex-dukhno , I could fix to let `thiserror` format the error message as below.

```
pub enum QueryError {
...
    #[error("column {0:?} does not exist")]
    ColumnDoesNotExist(Vec<String>),
...
```

The interaction error message will be:

```
xxx=> select column_not_in_table from schema_name.table_name;
ERROR:  column "["column_not_in_table"]" does not exist
```

I think it is not the right message you expected. Could you help give the template message of this error? Thanks.

**Comment by alex-dukhno** [comment]
Seems like I misunderstand again 😅 

`thiserror` formatting is used only in tests in `sql_engine` and user facing message is formatted here https://github.com/alex-dukhno/database/pull/74/files#diff-13fa0b4f67547213dbcccde5cd778787R165 🤦  😄 

so can you adjust user facing message
1) if single column to
```
ERROR:  column column_not_in_table does not exist
```
2) if more than one
```
ERROR:  columns column_1_not_in_table, column_1_not_in_table does not exist
```

thanks

**Comment by alex-dukhno** [comment]
You don't need to change latest changes for `QueryError`
```
    #[error("column {0:?} does not exist")]
    ColumnDoesNotExist(Vec<String>),
```
it is good for debugging in tests, Thanks

**Comment by silathdiir** [comment]
Hi @alex-dukhno , the CI has passed, please help review the latest code. Thanks.

File: src/node/src/node.rs [hunk 1]
```diff
@@ -307,6 +312,19 @@ mod mapper {
...
 
+    #[test]
+    fn column_does_not_exists() {
+        let non_existing_columns = vec!["column_not_in_table1".to_owned(), "column_not_in_table2".to_owned()];
+        assert_eq!(
+            QueryResultMapper::map(Err(QueryError::ColumnDoesNotExist(non_existing_columns.clone()))),
+            vec![Message::ErrorResponse(
+                Some("ERROR".to_owned()),
+                Some("42703".to_owned()),
+                Some(format!("column \"{:?}\" does not exist", non_existing_columns)),
```
Comment (alex-dukhno) [inline src/node/src/node.rs:323]:
```suggestion
                Some("columns column_not_in_table1, column_not_in_table2 does not exist".to_owned()),
```

File: src/node/src/node.rs [hunk 2]
```diff
@@ -307,6 +312,19 @@ mod mapper {
         )
     }
 
+    #[test]
+    fn column_does_not_exists() {
+        let non_existing_columns = vec!["column_not_in_table1".to_owned(), "column_not_in_table2".to_owned()];
```
Comment (alex-dukhno) [inline src/node/src/node.rs:317]:
```suggestion

```

File: src/node/src/node.rs [hunk 3]
```diff
@@ -307,6 +312,19 @@ mod mapper {
...
+        assert_eq!(
+            QueryResultMapper::map(Err(QueryError::ColumnDoesNotExist(non_existing_columns.clone()))),
```
Comment (alex-dukhno) [inline src/node/src/node.rs:319]:
```suggestion
            QueryResultMapper::map(Err(QueryError::ColumnDoesNotExist(vec!["column_not_in_table1".to_owned(), "column_not_in_table2".to_owned()]))),
```

File: src/node/src/node.rs [hunk 4]
```diff
@@ -159,6 +159,11 @@ impl QueryResultMapper {
                 Some("42P01".to_owned()),
                 Some(format!("table \"{}\" does not exist", table_name)),
             )],
+            Err(QueryError::ColumnDoesNotExist(non_existing_columns)) => vec![Message::ErrorResponse(
+                Some("ERROR".to_owned()),
+                Some("42703".to_owned()),
+                Some(format!("column \"{:?}\" does not exist", non_existing_columns)),
+            )],
```
Comment (alex-dukhno) [inline src/node/src/node.rs:166]:
```suggestion
            Err(QueryError::ColumnDoesNotExist(non_existing_columns)) => {
            let message = if non_existing_columns.len() > 1 {
                format!("columns {} do not exist, non_existing_columns.join(", "));
            } else {
                format!("column {}  does not exist", non_existing_columns[0]);
            };
            vec![Message::ErrorResponse(
                Some("ERROR".to_owned()),
                Some("42703".to_owned()),
                Some(message),
            )]
            },
```

**Comment by silathdiir** [comment]
Hi @alex-dukhno , sorry for my bad understanding. I fixed the code according to your suggestions. Please help review the code. Thanks.

File: src/node/src/node.rs [hunk 5]
```diff
@@ -159,6 +159,19 @@ impl QueryResultMapper {
                 Some("42P01".to_owned()),
                 Some(format!("table \"{}\" does not exist", table_name)),
             )],
+            Err(QueryError::ColumnDoesNotExist(non_existing_columns)) => {
+                let error_message = if non_existing_columns.len() > 1 {
+                    format!("columns {} does not exist", non_existing_columns.join(", "))
```
Comment (alex-dukhno) [inline src/node/src/node.rs:164]:
```suggestion
                    format!("columns {} do not exist", non_existing_columns.join(", "))
```

File: src/node/src/node.rs [hunk 6]
```diff
@@ -307,6 +320,35 @@ mod mapper {
...
+    fn multiple_columns_does_not_exists() {
+        assert_eq!(
+            QueryResultMapper::map(Err(QueryError::ColumnDoesNotExist(vec![
+                "column_not_in_table1".to_owned(),
+                "column_not_in_table2".to_owned()
+            ]))),
+            vec![Message::ErrorResponse(
+                Some("ERROR".to_owned()),
+                Some("42703".to_owned()),
+                Some("columns column_not_in_table1, column_not_in_table2 does not exist".to_owned()),
```
Comment (alex-dukhno) [inline src/node/src/node.rs:347]:
```suggestion
                Some("columns column_not_in_table1, column_not_in_table2 do not exist".to_owned()),
```

**Comment by alex-dukhno** [comment]
I added small suggestions and applied to PR and merged. Thanks for contribution

**Pull request merged**